pkg/components/prometheus-operator/component_conversion_test.go

```diff
@@ -0,0 +1,137 @@
+package prometheus //nolint:testpackage
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/hashicorp/hcl/v2"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	yamlserializer "k8s.io/apimachinery/pkg/runtime/serializer/yaml"
+	"k8s.io/client-go/util/jsonpath"
+
+	"github.com/kinvolk/lokomotive/pkg/components/util"
+)
+
+func renderManifests(t *testing.T, configHCL string) map[string]string {
+	name := "prometheus-operator"
+
+	component := newComponent()
+
+	body, diagnostics := util.GetComponentBody(configHCL, name)
+	if diagnostics.HasErrors() {
+		t.Fatalf("Getting component body: %v", diagnostics.Errs())
+	}
+
+	diagnostics = component.LoadConfig(body, &hcl.EvalContext{})
+	if diagnostics.HasErrors() {
+		t.Fatalf("Valid config should not return an error, got: %v", diagnostics)
+	}
+
+	ret, err := component.RenderManifests()
+	if err != nil {
+		t.Fatalf("Rendering manifests with valid config should succeed, got: %s", err)
+	}
+
+	return ret
+}
+
+//nolint:funlen
+func TestConversion(t *testing.T) {
+	testCases := []struct {
+		Name                   string
+		InputConfig            string
+		ExpectedConfigFileName string
+		Expected               reflect.Value
+		JSONPath               string
+	}{
+		{
+			Name:                   "verify foldersFromFilesStructure in configmap",
+			InputConfig:            `component "prometheus-operator" {}`,
+			ExpectedConfigFileName: "prometheus-operator/charts/grafana/templates/configmap-dashboard-provider.yaml",
+			Expected: reflect.ValueOf(`apiVersion: 1
+providers:
+- name: 'sidecarProvider'
+  orgId: 1
+  folder: ''
+  type: file
+  disableDeletion: false
+  allowUiUpdates: false
+  options:
+    foldersFromFilesStructure: true
+    path: /tmp/dashboards`),
+			JSONPath: "{.data.provider\\.yaml}",
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.Name, func(t *testing.T) {
+			t.Parallel()
+
+			m := renderManifests(t, tc.InputConfig)
+			if len(m) == 0 {
+				t.Fatalf("Rendered manifests shouldn't be empty")
+			}
+
+			gotConfig, ok := m[tc.ExpectedConfigFileName]
+			if !ok {
+				t.Fatalf("Config not found with filename: %q", tc.ExpectedConfigFileName)
+			}
+
+			u := getUnstructredObj(t, gotConfig)
+			got := getValFromObject(t, tc.JSONPath, u)
```

Comment:
It makes updating charts more difficult and less focused. IMO we shouldn't test all manifests, only the bits we care about.

Reply:
Don't we care about everything? The whole point of writing a test is to catch things we *didn't* expect.

Reply:
Hm, for me it's more about a regression test for features we enable via configuration in other parts of the system.

I think for catching this we didn't expect this provides little value, as usual workflow will be to just make those big YAMLs match, I doubt one will spend time to analyzing every single line changed. If we do that though, then upgrading components will be taking much more time and we should be aware of that.

Reply:
> Why not expect the entire YAML document and test against it rather than querying for specific fields? Looks simpler and more comprehensive to me.

I agree that it will simpler.

The test code will span some thousands of lines. And it will be a maintenance nightmare. There are parameters that have values of versions now these files will have to be updated in every component upgrade. With so huge a change in every upgrade we will miss details because large changes bring in "config fatigue" and erroneous code will fall through the cracks.

The path this test has taken is to make it more understandable and focused. We don't want to test upstream's work, but our changes on top of upstream.

Reply:
> Don't we care about everything? The whole point of writing a test is to catch things we didn't expect.

We should rather write focused tests even for such things. The problem with "things we didn't expect" analogy is that "any" upgrades/updates in config and parameter comes under this category.

We want to avoid "bad things we didn't expect" and IMHO focused tests will help us get there.

Reply:
I understand the motivation @surajssd. My concern was that the test looks rather complex the way it is now. But OK, we can evolve the test over time if we realize it's necessary.